cirq/sim/mps_simulator.py

```diff
@@ -0,0 +1,350 @@
+    ) -> Dict[str, List[np.ndarray]]:
+
+        param_resolver = param_resolver or study.ParamResolver({})
+        resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
+        self._check_all_resolved(resolved_circuit)
+
+        measurements = {}  # type: Dict[str, List[np.ndarray]]
+        if repetitions == 0:
+            for _, op, _ in resolved_circuit.findall_operations_with_gate_type(ops.MeasurementGate):
+                measurements[protocols.measurement_key(op)] = np.empty([0, 1])
+
+        for _ in range(repetitions):
+            all_step_results = self._base_iterator(
+                resolved_circuit, qubit_order=ops.QubitOrder.DEFAULT, initial_state=0
+            )
+
+            for step_result in all_step_results:
+                for k, v in step_result.measurements.items():
+                    if not k in measurements:
+                        measurements[k] = []
+                    measurements[k].append(np.array(v, dtype=int))
+
+        return {k: np.array(v) for k, v in measurements.items()}
+
+    def _check_all_resolved(self, circuit):
+        """Raises if the circuit contains unresolved symbols."""
+        if protocols.is_parameterized(circuit):
+            unresolved = [
+                op for moment in circuit for op in moment if protocols.is_parameterized(op)
+            ]
+            raise ValueError(
+                'Circuit contains ops whose symbols were not specified in '
+                'parameter sweep. Ops: {}'.format(unresolved)
+            )
+
+
+class MPSTrialResult(simulator.SimulationTrialResult):
+    def __init__(
+        self,
+        params: study.ParamResolver,
+        measurements: Dict[str, np.ndarray],
+        final_simulator_state: 'MPSState',
+    ) -> None:
+        super().__init__(
+            params=params, measurements=measurements, final_simulator_state=final_simulator_state
+        )
+
+        self.final_state = final_simulator_state
+
+    def __str__(self) -> str:
+        samples = super().__str__()
+        final = self._final_simulator_state
+        return f'measurements: {samples}\noutput state: {final}'
+
+
+class MPSSimulatorStepResult(simulator.StepResult):
+    """A `StepResult` that includes `StateVectorMixin` methods."""
+
+    def __init__(self, state, measurements):
+        """Results of a step of the simulator.
+        Attributes:
+            state: A MPSState
+            measurements: A dictionary from measurement gate key to measurement
+                results, ordered by the qubits that the measurement operates on.
+            qubit_map: A map from the Qubits in the Circuit to the the index
+                of this qubit for a canonical ordering. This canonical ordering
+                is used to define the state vector (see the state_vector()
+                method).
+        """
+        self.measurements = measurements
+        self.state = state
```

Comment:
If this result represents an intermediate state of a circuit, could mutating the `state` of this object affect subsequent steps in the simulation? Compare PR #3109 and linked issues.

Reply:
I am not sure, actually. I was just inspiring myself from this existing code:
https://sourcegraph.com/github.com/quantumlib/Cirq@9bad962c32a236ba72d829e5091d3bbc7711814c/-/blob/cirq/sim/clifford/clifford_simulator.py#L360

Which of course, could be wrong, or (more likely) I don't understand what I'm doing.

Leaving unresolved.

Reply:
I think the best path to resolving this is to add a test similar to the one in #3109, where you attempt to mutate an intermediate state and confirm that the final state isn't also mutated. If you find that it is mutated, `CliffordSimulator` should be updated as well.

Given that this touches code outside the scope of this PR, I'd also be fine with this being addressed in a follow-up PR.

Reply:
There was indeed a problem. I think that the Clifford simulator also had the same issue so instead of a TODO, I decided to send you another PR: #3664